aws/resource_aws_dlm_lifecycle_policy.go

```diff
@@ -0,0 +1,344 @@
+													ValidateFunc: validation.StringInSlice([]string{
+														dlm.IntervalUnitValuesHours,
+													}, false),
+												},
+												"times": {
+													Type:     schema.TypeList,
+													Optional: true,
+													Computed: true,
+													MaxItems: 1,
+													Elem: &schema.Schema{
+														Type:         schema.TypeString,
+														ValidateFunc: validation.StringMatch(regexp.MustCompile("^([0-9]|0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$"), "see https://docs.aws.amazon.com/dlm/latest/APIReference/API_CreateRule.html#dlm-Type-CreateRule-Times"),
+													},
+												},
+											},
+										},
+									},
+									"name": {
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: validation.StringLenBetween(0, 500),
+									},
+									"retain_rule": {
+										Type:     schema.TypeList,
+										Required: true,
+										MaxItems: 1,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"count": {
+													Type:         schema.TypeInt,
+													Required:     true,
+													ValidateFunc: validation.IntBetween(1, 1000),
+												},
+											},
+										},
+									},
+									"tags_to_add": {
+										Type:     schema.TypeMap,
+										Optional: true,
+									},
+								},
+							},
+						},
+						"target_tags": {
+							Type:     schema.TypeMap,
+							Required: true,
+						},
+					},
+				},
+			},
+			"state": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  dlm.SettablePolicyStateValuesEnabled,
+				ValidateFunc: validation.StringInSlice([]string{
+					dlm.SettablePolicyStateValuesDisabled,
+					dlm.SettablePolicyStateValuesEnabled,
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceAwsDlmLifecyclePolicyCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dlmconn
+
+	input := dlm.CreateLifecyclePolicyInput{
+		Description:      aws.String(d.Get("description").(string)),
+		ExecutionRoleArn: aws.String(d.Get("execution_role_arn").(string)),
+		PolicyDetails:    expandDlmPolicyDetails(d.Get("policy_details").([]interface{})),
+		State:            aws.String(d.Get("state").(string)),
+	}
+
+	log.Printf("[INFO] Creating DLM lifecycle policy: %s", input)
+	out, err := conn.CreateLifecyclePolicy(&input)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*out.PolicyId)
+
+	return resourceAwsDlmLifecyclePolicyRead(d, meta)
+}
+
+func resourceAwsDlmLifecyclePolicyRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dlmconn
+
+	log.Printf("[INFO] Reading DLM lifecycle policy: %s", d.Id())
+	out, err := conn.GetLifecyclePolicy(&dlm.GetLifecyclePolicyInput{
+		PolicyId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ResourceNotFoundException" && !d.IsNewResource() {
+			d.SetId("")
```

Comment:
Nit: When triggering resource recreation, we should return a warning message in the logs so operators and code maintainers can more easily diagnose why it occurred, e.g.

```go
log.Printf("[WARN] DLM Lifecycle Policy (%s) not found, removing from state", d.Id())
d.SetId("")
```